File: medium/Add_Two_Numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        l1_str = ''
        while l1.next is not None:
            l1_str = str(l1.val) + l1_str
            l1 = l1.next
        l1_str = str(l1.val) + l1_str
        
        l2_str = ''
        while l2.next is not None:
            l2_str = str(l2.val) + l2_str
            l2 = l2.next
        l2_str = str(l2.val) + l2_str
        
        l3_str = str(int(l1_str) + int(l2_str))
        l3 = ListNode(int(l3_str[0]))
        for i in l3_str[1:]:
            last_head_val = l3.val
            last_head_next=l3.next
            l3.val=i
            l3.next=ListNode()
            l3.next.val=last_head_val
            l3.next.next=last_head_next
        
        return l3

File: cpp/medium/Add_Two_Numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int overflow = 0;
        int sum = l1->val + l2->val + overflow;
        if (sum >= 10) {
            overflow = 1;
            sum -= 10;
        }
        else overflow = 0;
        ListNode* answer = new ListNode(sum);
        ListNode* next_answer = answer;
        while (l1->next or l2->next) {
            if (l1->next and l2->next) {
                l1 = l1->next;
                l2 = l2->next;
                sum = l1->val + l2->val + overflow;
            }
            else if (l1->next) {
                l1 = l1->next;
                sum = l1->val + overflow;
            }
            else {
                l2 = l2->next;
                sum = l2->val + overflow;
            }
            if (sum >= 10) {
                overflow = 1;
                sum -= 10;
            }
            else overflow = 0;
            next_answer->next = new ListNode(sum);
            next_answer = next_answer->next;
        }
        if (overflow) {
            next_answer->next = new ListNode(overflow);
        }
        return answer;
    }
};